plugins/tabletools/plugin.js

```diff
@@ -531,37 +552,41 @@
 			oCol.push( row[ colIndex ] );
 
 			// Avoid adding duplicate cells.
-			if ( row[ colIndex ].rowSpan > 1 )
+			if ( row[ colIndex ].rowSpan > 1 ) {
 				r += row[ colIndex ].rowSpan - 1;
+			}
 		}
 		return oCol;
 	}
 
 	function mergeCells( selection, mergeDirection, isDetect ) {
-		var cells = getSelectedCells( selection );
+		var cells = getSelectedCells( selection ),
+			commonAncestor;
 
 		// Invalid merge request if:
 		// 1. In batch mode despite that less than two selected.
 		// 2. In solo mode while not exactly only one selected.
 		// 3. Cells distributed in different table groups (e.g. from both thead and tbody).
-		var commonAncestor;
-		if ( ( mergeDirection ? cells.length != 1 : cells.length < 2 ) || ( commonAncestor = selection.getCommonAncestor() ) && commonAncestor.type == CKEDITOR.NODE_ELEMENT && commonAncestor.is( 'table' ) )
+		if ( ( mergeDirection ? cells.length != 1 : cells.length < 2 ) ||
+			( commonAncestor = selection.getCommonAncestor() ) &&
```

Comment:
I looked at this conditional statement a bit closer and it seems a bit weird... Normally I'd say that extracting this may be risky, because if `( mergeDirection ? cells.length != 1 : cells.length < 2 )` is true, the part after `||` (so with `commonAncestor`) will not be calculated and maybe it shouldn't because of some possible bugs. But... later we have `commonAncestor.type == CKEDITOR.NODE_ELEMENT` and `commonAncestor.is( 'table' )` that should throw `TypeError` if `( commonAncestor = selection.getCommonAncestor() )` didn't run, so it seems that the first condition is always false 🤔 So after all I think your suggestion may actually save us some super-duper-edge-case error in the future 😂 (of course please correct me if I'm wrong).

Reply:
So, we already see  that this code is unreadable as hell :smile: Look how it splits:

```javascript
var isBatchMode = mergeDirection ? cells.length != 1 : cells.length < 2;

if ( isBatchMode ) {
  return false;
}

var commonAncestor = selection.getCommonAncestor();

if ( isTableElement( commonAncestor ) ) {
  return false;
}

function isTableElement( node ) {
  return node && node.type === CKEDITOR.NODE_ELEMENT && node.is( 'table' );
}
```

It's the same code as this crazy `if` - although please, make sure to test it and verify if it's correct, I could also make some mistake.

Reply:
About the case with:
> But... later we have commonAncestor.type == CKEDITOR.NODE_ELEMENT and commonAncestor.is( 'table' ) that should throw TypeError if ( commonAncestor = selection.getCommonAncestor() ) didn't run, so it seems that the first condition is always false thinking
look on brackets here: `( commonAncestor = selection.getCommonAncestor() ) && ...`, code after && won't be evaulated if `commonAncestor` result with `null`.

Reply:
I think the first early return is unsafe - we can simplify the original `if` (overindentation is deliberate 😛): 
```javascript
if ( 
    ( mergeDirection ? cells.length != 1 : cells.length < 2 ) || 
    ( commonAncestor = selection.getCommonAncestor() ) && 
    commonAncestor.type == CKEDITOR.NODE_ELEMENT && 
    commonAncestor.is( 'table' ) 
    ) {
    return false;
}
```
to:
```javascript
if ( a || b && c && d )
```
So even if `a == true` we still need `c == true` and `d == true`.
But... `c` and `d` can only be `true` if `b == true` (or we get TypeError as I mentioned before). In fact, `a == false` doesn't change anything if all `b, c, d` are equal to `true` 😱So it seems to me that all that code can be replaced with just:

```javascript
var commonAncestor = selection.getCommonAncestor(),
    ancestorIsNode = commonAncestor && ( commonAncestor.type == 
    CKEDITOR.NODE_ELEMENT ),
    ancestorIsTable = ancestorIsNode && commonAncestor.is( 'table' );

if ( ancestorIsTable ) {
    return;
}
```

Reply:
Let's slow down by removing the first condition :smile: 

```javascript
if ( 
    ( mergeDirection ? cells.length != 1 : cells.length < 2 ) ||  // If true, the rest of the code won't evaluate and function returns false. If false, we are going deeper.
    ( commonAncestor = selection.getCommonAncestor() ) && // If null, the rest of the code won't evaluate. If node, going deeper.
    commonAncestor.type == CKEDITOR.NODE_ELEMENT &&  // If false, nothing. True - deeper.
    commonAncestor.is( 'table' ) // False, nothing. True - function returns false. 
    ) {
    return false;
}
```

However, if you are still not sure about this refactoring, let's leave it as, so we won't accidentally introduce any bug.

Reply:
Yeah, you are totally right, I misanalysed this... Still as this piece of code isn't covered by tests (removing any condition or even entire `if` statement doesn't cause any test to fail) it is probably best to leave it as is to merge this PR as quickly as possible.